### ISSUE TITLE
Using Python OOP with Eel

### ISSUE THREAD

**Issue by KilliKrate** [ChrisKnott/Eel#141]
I would like to use a class structure inside my python file, but if i try to expose a class method, it says that "self" is missing. Using objects for my project is extremely important and since I haven't found any examples using classes, I thought it would be important to know: is there a way to expose and use python class methods from inside a js file?

**Comment by ChrisKnott** [comment]
There's not really any way of doing this, because there's no way of know what object the method is being called on. You could expose them with dedicated methods like...
```@eel.expose
def my_method():
  whatever_object.method()
```
Otherwise, how would you know which object the method is being called on?

The only other way I can think of doing this is to keep a map of objects by their id;
```objs = {id(o): o for o in objects}```

Then you can pass the id where you would normally have the 'self' reference. Something like this...
```
class MyClass:
  def my_method(self):
    ...

@eel.expose
def my_method(self_id):
  objs[self_id].my_method()
```

----

Basically, there's not an easy way of doing it because your objects don't exist as full Python objects within the Javascript interpreter.

**Comment by KilliKrate** [comment]
Ok, I understand. Thanks for your time!

**Comment by KilliKrate** [comment]
Whoops, before I go, i just wanted to tell you that it would truly be amazing if you could use objects with this library. I managed to get my project working with objects thanks to the pywebview library, but it's kind of awful because it runs on an integrated IE11 renderer and because of that you cannot use async/await, or any ES6 and ES7 features. I would totally use Eel over pywebview, if this feature was available. Thanks again!

**Comment by IASV** [comment]
Hi! a solution fast. Create a **Class** and exposed the class method that do you need to use.

```python
import eel

class Main:
	def __init__(self):
		self.eel = eel
		self.eel.init("templates")
		self.name = 'test'
	
	def print_name(self):
		print(self.name)
	
	def start_ui(self):
		self.eel.start("index.html", size(500, 500), position(200,200))

if __name__ == "__main__":
	app = Main()
	eel.expose(app.print_name)
	app.start_ui()

```

**Comment by AurelienDellac** [comment]
Hi @IASV. I do not really understand your answer. How do you expose a class method with the "self" argument ?

**Comment by alexandremassoutier** [comment]
Hi, I know this topic is quite old but actually working on a project using EEL PYTHON (congrats to the creator and thanks), I found an easy and straight solution to KilliKrate demand.

I reformulate.

1) Using POO in Python will obviously force some of our EEL expose functions within classes (app, gui, config...) to have self parameter as 1st argument.

2) When calling an exposed function from a JS file, "self" cannot be known unless Python sends it and retreives it in the same shot. Otherwise error thrown in Python will be "missing argument".

3) Simple solution is to init on JS references to our classes that will have to deal with the web view. I think 2 or 3 max. In my project I sent to js my app class, my gui and cfg class and for now it is ok.

4) Process = create for instance the following function on JS side :
// Python main objects
let app;
let cfg;
let gui;
/* Function that sets main python objects */
eel.expose(setPythonMainObjects);
function setPythonMainObjects(name, obj) {
  switch(name) {
    case 'app':
      app = obj;
      break;
    case 'cfg':
      cfg = obj;
      break;
    case 'gui':
      gui = obj;
      break;
    default:
      console.log("FATAL - Main Python Object " + name + " could not be set");
  }
}

On Python side, when all objects are instanciated call the function sending app, gui, etc...

This works.

**Comment by ncl-a** [comment]
Hi (I'm new), my solution is:

```python
import eel


class MyClass:
   def my_method(self):
      pass
      
my_class = MyClass()
eel._expose("other_name", my_class.my_method)
```

**Comment by alexandremassoutier** [comment]
Hi,
I appreciate you took the time to answer but I could not make it work.
Maybe I did something wrong.

Basically, to make my situation clear :
I've a app.py file, the entry point.I init a App class which inits eel by doing eel.init(...) so that everyting in web directory be disposable as quick as possible.Then app class inits Gui class and in its Run method launches Gui.Run which starts eel.
For now my @eel.expose methods can only exists in my App class in the main Python environment (top level).

What I would like for a better architecture :
As a strat, a EelCatalog class where I could set all methods my JS will be dealing with.
---

Again I thank for having answered something.
Alexandre

 

    Le vendredi 31 mars 2023 à 18:41:36 UTC+2, ncla ***@***.***> a écrit :  
 
 


Hi (I'm new), my solution is:
import eel


class MyClass:
   def my_method(self):
      pass
      
my_class = MyClass()
eel._expose("other_name", my_class.my_method)

**Comment by LukasMandok** [comment]
> Hi! a solution fast. Create a **Class** and exposed the class method that do you need to use.
> 
> ```python
> import eel
> 
> class Main:
> 	def __init__(self):
> 		self.eel = eel
> 		self.eel.init("templates")
> 		self.name = 'test'
> 	
> 	def print_name(self):
> 		print(self.name)
> 	
> 	def start_ui(self):
> 		self.eel.start("index.html", size(500, 500), position(200,200))
> 
> if __name__ == "__main__":
> 	app = Main()
> 	eel.expose(app.print_name)
> 	app.start_ui()
> ```

Thanks, this approach works for me as well, my code looks something like this:

```python
import eel
import os

class Application:
    def __init__(self):
        self.html_dir  = os.path.dirname(__file__)
        self.html_file = os.path.join(self.html_dir, 'webpage', 'index.html')

        self.eel = eel

    def run(self):
        self.eel.init(self.html_dir)
        self.eel.expose(self.example_function)
        self.eel.start(self.html_file, size=(500, 500), mode='firefox')

    def example_function(self, arg1):
        print(arg1)
```
